File: algorithms/algorithms-java/src/main/java/com/lanceolata/coding/algorithms/java/leetcode/Question_1019_Next_Greater_Node_In_Linked_List.java
package com.lanceolata.coding.algorithms.java.leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class Question_1019_Next_Greater_Node_In_Linked_List {
    // Definition for singly-linked list.
    public class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }

    public int[] nextLargerNodes(ListNode head) {
        List<Integer> res = new ArrayList<>();
        Stack<Integer> stack = new Stack<>();
        ListNode p = head;
        while (p != null) {
            while (!stack.isEmpty() && res.get(stack.peek()) < p.val) {
                res.set(stack.pop(), p.val);
            }
            stack.push(res.size());
            res.add(p.val);
            p = p.next;
        }
        for (int i : stack) {
            res.set(i, 0);
        }
        int[] arr = new int[res.size()];
        for (int i = 0; i < arr.length; i++) {
            arr[i] = res.get(i);
        }
        return arr;
    }
}

File: algorithms/algorithms-cplusplus/leetcode/Question_1019_Next_Greater_Node_In_Linked_List.cc
#include <vector>

using namespace std;

// Definition for singly-linked list.
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
 public:
  vector<int> nextLargerNodes(ListNode* head) {
    vector<int> res, stack;
    ListNode* p = head;
    while (p) {
      while (stack.size() && res[stack.back()] < p->val) {
        res[stack.back()] = p->val;
        stack.pop_back();
      }
      stack.push_back(res.size());
      res.push_back(p->val);
      p = p->next;
    }
    for (int i: stack) {
      res[i] = 0;
    }
    return res;
  }
};
